src/cpu/flexcpu/flexcpu_thread.cc

```diff
@@ -0,0 +1,1673 @@
+
+        delete resp;
+
+        if (f != NoFault) {
+            markFault(inst_ptr, f);
+            return;
+        }
+
+        // NOTE: We specifically capture the shared_ptr instead of a weak_ptr
+        //       in this case, because stores are committed once sent, and need
+        //       to be kept alive long enough for the completion event to
+        //       occur. Therefore after this point, the lifespan of the
+        //       InflightInst is managed by this lambda released into the wild
+        //       instead of in the inflightInsts buffer.
+        auto callback =
+            [this, inst_ptr] (Fault fault) {
+                onExecutionCompleted(inst_ptr, fault);
+            };
+
+        if (sreq) { // split
+            assert(sreq->high && sreq->low);
+            _cpuPtr->requestSplitMemWrite(sreq->main, sreq->low,
+                sreq->high, getThreadContext(),
+                inst_ptr->staticInst(),
+                static_pointer_cast<ExecContext>(inst_ptr),
+                inst_ptr->traceData(), data.get(),
+                callback);
+        } else { // not split
+            _cpuPtr->requestMemWrite(req, getThreadContext(),
+                inst_ptr->staticInst(),
+                static_pointer_cast<ExecContext>(inst_ptr),
+                inst_ptr->traceData(), data.get(), callback);
+        }
+
+        commitAllAvailableInstructions();
+    } else { // read
+        auto callback =
+            [this, weak_inst] (Fault fault) {
+                onExecutionCompleted(weak_inst, fault);
+            };
+        if (sreq) { // split
+            assert(sreq->high && sreq->low);
+            _cpuPtr->requestSplitMemRead(sreq->main, sreq->low,
+                sreq->high, getThreadContext(),
+                inst_ptr->staticInst(),
+                static_pointer_cast<ExecContext>(inst_ptr),
+                inst_ptr->traceData(),
+                callback);
+        } else { // not split
+            _cpuPtr->requestMemRead(req, getThreadContext(),
+                inst_ptr->staticInst(),
+                static_pointer_cast<ExecContext>(inst_ptr),
+                inst_ptr->traceData(), callback);
+        }
+    }
+}
+
+void
+FlexCPUThread::squashUpTo(InflightInst* const inst_ptr, bool rebuild_lasts)
+{
+    bool last_ser_correct = true;
+    bool last_mem_bar_correct = true;
+
+    shared_ptr<InflightInst> ser_inst = lastSerializingInstruction.lock();
+    shared_ptr<InflightInst> bar_inst = lastMemBarrier.lock();
+    size_t count = 0;
+    while (!inflightInsts.empty() && inflightInsts.back().get() != inst_ptr) {
+        const shared_ptr<InflightInst>& back_inst = inflightInsts.back();
+        back_inst->notifySquashed();
+        if (back_inst == ser_inst) last_ser_correct = false;
+        if (back_inst == bar_inst) last_mem_bar_correct = false;
+        recordInstStats(back_inst);
+        inflightInsts.pop_back();
+        ++count;
+    }
+
```

Comment:
Add assert that we've squashed something

Reply:
Actually, with branch prediction, it could be the instruction at the tail of the buffer where we call "squashUpTo" on, which would result in 0 things being squashed. This assert would be invalid in that case.